### DISCUSSION TITLE
Scenario for automatic fallback mechanism

### DISCUSSION THREAD

**Discussion by Unknown78** [pal1000/mesa-dist-win#173] [Ideas]
Based on

> https://github.com/pal1000/mesa-dist-win/discussions/171#discussioncomment-8554152
> There is no automatic fallback mechanism unfortunately otherwise I wouldn't have developed a per app deployment tool.

So this is my idea,

If an App hardware acceleration settings enabled then default GPU render > Mesa GPU render > Mesa CPU Render
If an App hardware acceleration settings disabled then default CPU render > Mesa CPU Render

The fallback should exist for each component of compliancy.

<hr>

For example in a system with only 70% OpenGL 3.3 support,

If an app hardware acceleration settings are enabled and it needs 100% OpenGL 3.3: 
The 70% OpenGL 3.3 will be default GPU render, then the rest 30% would be Mesa GPU render, if that fail then 30% Mesa CPU render.

If an app hardware acceleration settings are disabled and it needs 100% OpenGL 3.3:
The 70% OpenGL 3.3 will be default CPU render, then the rest 30% must be Mesa CPU render.

<hr>

Is there anything I miss?

**Comment by pal1000** [comment]
Automatic fallback would require an user-mode OpenGL driver that would handle CPU lowering from GPU to Mesa llvmpipe and alter registry so GPU keernel driver to load that driver.

**Reply by Unknown78** [reply]
So it acts as a main driver for the proxy of all other graphic drivers?

**Reply by pal1000** [reply]
Yes we need a proxy driver to handle automatic fallback.
